internal/exec/util/blkid.c

```diff
@@ -141,10 +141,26 @@ result_t blkid_get_partition(const char *device, int part_num, struct partition_
 	if (!part)
 		return RESULT_BAD_INDEX;
 
-	return extract_part_info(part, info);
+	blkid_topology topo = blkid_probe_get_topology(pr);
+	if (!topo) {
+		return RESULT_NO_TOPO;
+	}
+
+	long sector_size = blkid_topology_get_logical_sector_size(topo);
+	if (!sector_size) {
+		return RESULT_NO_SECTOR_SIZE;
+	}
+	if (sector_size % 512 != 0) {
+		return RESULT_BAD_SECTOR_SIZE;
+	}
+
+	return extract_part_info(part, info, sector_size / 512);
 }
 
-static result_t extract_part_info(blkid_partition part, struct partition_info *info)
+// extract_part_info reads the information for a partition into *info. sector_divisor is how many 512
+// byte sectors are in a logical sector (1 for "normal" sectors, 8 for 4k sectors). This is needed because
+// libblkid always assumes 512 byte sectors regardless of what the actual logical sector size of the device is.
+static result_t extract_part_info(blkid_partition part, struct partition_info *info, long sector_divisor)
```

Comment:
It sounds like the last parameter is just a boolean for `is_4k`. As I don't think there are other options outside of 512/4096, perhaps it may make sense to check for those two cases in the caller and just have a boolean here? (just wondering, not a strong opinion).

Reply:
I don't see a reason to not support other cases, especially since it's no more work. I think losetup allows arbitrary power of two.